akka-stream/src/main/scala/akka/stream/remote/scaladsl/SinkRef.scala

```diff
@@ -0,0 +1,199 @@
+/**
+ * Copyright (C) 2014-2017 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.remote.scaladsl
+
+import akka.actor.{ Actor, ActorLogging, ActorRef, Props, Terminated }
+import akka.annotation.InternalApi
+import akka.pattern.FutureRef
+import akka.stream._
+import akka.stream.impl.SinkModule
+import akka.stream.remote.impl.StreamRefsMaster
+import akka.stream.remote.{ RemoteStreamRefActorTerminatedException, StreamRefs }
+import akka.stream.scaladsl.Source
+import akka.stream.stage._
+import akka.util.{ ByteString, OptionVal }
+import org.reactivestreams.Subscriber
+
+import scala.concurrent.{ Future, Promise }
+
+object SinkRef {
+  def source[T](): Source[T, Future[SinkRef[T]]] =
+    Source.fromGraph(new SinkRefLocalImpl[T]()) // TODO settings?
+
+  def bulkTransferSource(port: Int = -1): Source[ByteString, SinkRef[ByteString]] = {
+    ???
+  }
+}
+
+///**
+// * INTERNAL API
+// */
+//@InternalApi private[akka] final class SinkRefModule[In](val attributes: Attributes, shape: SinkShape[In]) extends SinkModule[In, ActorRef](shape) {
+//
+//  override def create(context: MaterializationContext): (Subscriber[In], ActorRef) = {
+//    val system = ActorMaterializerHelper.downcast(context.materializer)
+//    val subscriberRef = system.actorOf(context.copy(islandName = ""), SinkRefActor.props)
+//    (akka.stream.actor.ActorSubscriber[In](subscriberRef), subscriberRef)
+//  }
+//
+//  override protected def newInstance(shape: SinkShape[In]): SinkModule[In, ActorRef] = new SinkRefModule[In](attributes, shape)
+//  override def withAttributes(attr: Attributes): SinkModule[In, ActorRef] = new SinkRefModule[In](attr, amendShape(attr))
+//}
+
+final class SinkRefLocalImpl[T] extends GraphStageWithMaterializedValue[SourceShape[T], Future[SinkRef[T]]] {
+  final val out: Outlet[T] = Outlet[T]("SinkRefSource.out")
+  override final def shape = SourceShape.of(out)
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes) = {
+    val promise = Promise[SinkRef[T]]()
+
+    /*
+     * This stage can only handle a single "sender" (it does not merge values);
+     * The first that pushes is assumed the one we are to trust
```

Comment:
We could also add a handshake, where a potential sender will make sure it is the first.

Reply:
I'm also thinking that we can carry the "initial demand" in the SinkRef that we send over to the other side. One configures it locally, and during serialization, the demand value gets carried along with it. :)